torch/csrc/jit/passes/freeze_module.cpp

```diff
@@ -318,55 +291,151 @@ class AttributePropagator {
+              }
+              break;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // This function recursively iterates over submodules to identify
+  // for each class type the attribute slots that need to be preserved.
+  //
+  // Note 'attrsToKeep[type].insert(type->numAttributes())' means all
+  // attribute slots of 'type' and its methods are preserved. A submodule is
+  // preserved when it escapes (meaning it is returned).
+  void handleSharedClassType(Module& module, std::shared_ptr<Graph>& graph) {
+    auto type = module.type();
+    size_t N = type->numAttributes();
+    if (moduleEscapes(module, graph)) {
+      // Perserve all its attributes and methods.
+      attrsToKeep_[type].insert(N);
+      return;
+    }
+    auto it2 = preservedScalarAttrs_.find(module._ivalue());
+    SharedTypeSubModules_[type].insert(module._ivalue());
+    attrsToKeep_[type].insert({});
     for (size_t i = 0; i < N; ++i) {
       auto attrTy = type->getAttribute(i);
       auto name = type->getAttributeName(i);
-      auto attr = module_.attr(name);
-      bool immutable;
+      auto attr = module.attr(name);
+      bool isMutable;
       if (AliasDb::mutableType(attrTy)) {
-        immutable = it == preservedAttrs_.end() || !it->second.count(attr);
+        isMutable = preservedAttrs_.count(attr);
       } else {
-        immutable =
-            it2 == preservedScalarAttrs_.end() || !it2->second.count(name);
+        isMutable =
+            it2 != preservedScalarAttrs_.end() && it2->second.count(name);
       }
-      if (immutable) {
-        attrsToRemove.push_back(name);
+      if (isMutable) {
+        attrsToKeep_[type].insert(i);
+        if (attr.isModule()) {
+          auto attrModule = attr.toModule();
+          handleSharedClassType(attrModule, graph);
+        }
       }
     }
-    for (auto& name : attrsToRemove) {
-      module_._ivalue()->unsafeRemoveAttr(name);
-      module_.type()->unsafeRemoveAttribute(name);
-    }
-    for (auto& fn : type->methods()) {
-      auto& name = fn->name();
-      if ("forward" == name)
+  }
+
+  // Remove unused attributes and methods for each sub module of the frozen
+  // module. This function iterates over the Calsstypes of its submodule
+  // attributes including its own type.
+  void removeUnusedAttrs() {
+    std::vector<std::string> attrsToRemove;
+    std::vector<Function*> funcsToRemove;
+    for (auto& it : attrsToKeep_) {
+      auto& type = it.first;
+      size_t N = type->numAttributes();
+      if (it.second.count(N)) {
         continue;
-      type->unsafeRemoveMethod(name);
-      module_._ivalue()->compilation_unit()->unsafeRemoveMethod(fn->qualname());
+      }
+      for (size_t i = 0; i < N; ++i) {
+        if (it.second.count(i) == 0) {
+          attrsToRemove.push_back(type->getAttributeName(i));
+        }
+      }
+      for (auto& fn : type->methods()) {
+        auto& name = fn->name();
+        if ("forward" == name && type == module_.type())
```

Comment:
In this case you may have the same type ptr but probably safer to do `*type == module_.type()`